### PR TITLE
Add MFC 1/4K 4/7bUID to "Add Manually"

### PR DESCRIPTION
**Pull request by Astrrra**
# What's new

- The "Add Manually" section in the NFC app now allows you to create blank Mifare Classic 1K/4K cards with either a 4-byte or a 7-byte UID.

# Verification 

- Create each of the 4 card variations. Verify that the internal data structure matches the datasheet (UID in the block 0, all access conditions set to FF0780 and all keys set to FFFFFFFFFFFF as per section 8.6.3)

# Checklist (For Reviewer)

- [x] PR has description of feature/bug or link to Confluence/Jira task
- [x] Description contains actions to verify feature/bugfix
- [x] I've built this code, uploaded it to the device and verified feature/bugfix

### CODE REVIEW

File: applications/nfc/helpers/nfc_generators.c [hunk 1]
```diff
@@ -333,6 +460,26 @@ static const NfcGenerator ntag_i2c_plus_2k_generator = {
...
+
+static const NfcGenerator mifare_classic_4k_4b_uid_generator = {
+    .name = "Mifare Classic 4k 4byte UID",
+    .generator_func = nfc_generate_mf_classic_4k_4b_uid,
+    .next_scene = NfcSceneMfClassicMenu};
+
+static const NfcGenerator mifare_classic_4k_7b_uid_generator = {
+    .name = "Mifare Classic 4k 7byte UID",
+    .generator_func = nfc_generate_mf_classic_4k_7b_uid,
+    .next_scene = NfcSceneMfClassicMenu};
```
Comment (DrZlo13) [inline applications/nfc/helpers/nfc_generators.c]: Please add a comma to the last line of each generator to make it more readable.

**Review by DrZlo13 — CHANGES_REQUESTED** [review]

File: applications/nfc/helpers/nfc_generators.c (hunk 1 above)
Comment (Astrrra) [reply applications/nfc/helpers/nfc_generators.c]: Fixed

**Comment by skotopes** [comment]
Sync with dev please

File: applications/nfc/helpers/nfc_generators.c [hunk 2]
```diff
@@ -25,6 +25,39 @@ static void nfc_generate_mf_ul_uid(uint8_t* uid) {
...
+    furi_assert(block != NULL);
+    nfc_generate_mf_classic_uid(block, uid_len);
+    for(int i = uid_len; i < 16; i++) {
+        block[i] = 0xFF;
+    }
+}
+
+static void nfc_generate_mf_classic_sector_trailer(MfClassicData* data, uint8_t block) {
+    // All keys are set to FFFF FFFF FFFFh at chip delivery and the bytes 6, 7 and 8 are set to FF0780h.
+    data->block[block].value[6] = 0xFF;
```
Comment (gornekich) [inline applications/nfc/helpers/nfc_generators.c]: Please rework with MfClassicSectorTrailer customization

File: applications/nfc/helpers/nfc_generators.c [hunk 3]
```diff
@@ -25,6 +25,39 @@ static void nfc_generate_mf_ul_uid(uint8_t* uid) {
...
+    data->block[block].value[7] = 0x07;
+    data->block[block].value[8] = 0x80;
+    data->block[block].value[9] = 0x69; // Nice
+    for(int i = 0; i < 16; i++) {
+        if(i < 6 || i > 9) {
+            data->block[block].value[i] = 0xFF;
+        }
+    }
```
Comment (gornekich) [inline applications/nfc/helpers/nfc_generators.c]: memset(sec_trailer->key_a, 0xff, sizeof(sec_trailer->key_a))

File: applications/nfc/helpers/nfc_generators.c (hunk 2 above)
Comment (gornekich) [reply applications/nfc/helpers/nfc_generators.c]: MfClassicSectorTrailer* sec_tr = (MfClassicSectorTrailer* ) data->block[block].value
sec_tr->access_bits  = ...

File: applications/nfc/helpers/nfc_generators.c [hunk 4]
```diff
@@ -25,6 +25,39 @@ static void nfc_generate_mf_ul_uid(uint8_t* uid) {
...
 
+static void nfc_generate_mf_classic_uid(uint8_t* uid, uint8_t length) {
+    uid[0] = NXP_MANUFACTURER_ID;
+    furi_hal_random_fill_buf(&uid[1], length - 1);
+}
+
+static void nfc_generate_mf_classic_block_0(uint8_t* block, uint8_t uid_len) {
+    // Block length is always 16 bytes, and the UID can be either 4 or 7 bytes
+    furi_assert(uid_len == 4 || uid_len == 7);
+    furi_assert(block != NULL);
```
Comment (gornekich) [inline applications/nfc/helpers/nfc_generators.c]: furi_assert(block) is better for me

File: applications/nfc/helpers/nfc_generators.c [hunk 5]
```diff
@@ -268,70 +314,188 @@ static void nfc_generate_ntag_i2c_plus_2k(NfcDeviceData* data) {
...
+        data->nfc_data.uid[i] = data->mf_classic_data.block[0].value[i];
+    }
+
+    MfClassicData* mfc = &data->mf_classic_data;
+    mf_classic_set_block_read(mfc, 0, &mfc->block[0]);
+
+    // Set every block to 0xFF
+    for(uint16_t i = 1; i < MF_CLASSIC_1K_TOTAL_SECTORS_NUM * 4; i += 1) {
+        memset(&mfc->block[i].value, 0xFF, 16);
+        mf_classic_set_block_read(mfc, i, &mfc->block[i]);
```
Comment (gornekich) [inline applications/nfc/helpers/nfc_generators.c]: Next for cycle is not necessary
if(mf_classic_is_sector_trailer(i)) {
    nfc_generate_mf_classic_sector_trailer(mfc, i);
 } else {
    memset(&mfc->block[i].value, 0xFF, 16);
    }
mf_classic_set_block_read(mfc, i, &mfc->block[i]);

File: applications/nfc/helpers/nfc_generators.c [hunk 6]
```diff
@@ -268,70 +314,188 @@ static void nfc_generate_ntag_i2c_plus_2k(NfcDeviceData* data) {
...
+    }
+
+    // Set every 4th block with the sector trailer
+    for(uint16_t i = 3; i < MF_CLASSIC_1K_TOTAL_SECTORS_NUM * 4; i += 4) {
```
Comment (gornekich) [inline applications/nfc/helpers/nfc_generators.c]: Unnecessary

File: applications/nfc/helpers/nfc_generators.c [hunk 7]
```diff
@@ -268,70 +314,188 @@ static void nfc_generate_ntag_i2c_plus_2k(NfcDeviceData* data) {
...
+
+static void nfc_generate_mf_classic_1k_4b_uid(NfcDeviceData* data) {
+    nfc_generate_mf_classic_1k(data, 4);
+}
+
+static void nfc_generate_mf_classic_1k_7b_uid(NfcDeviceData* data) {
+    nfc_generate_mf_classic_1k(data, 7);
+}
+
+static void nfc_generate_mf_classic_4k(NfcDeviceData* data, uint8_t uid_len) {
```
Comment (gornekich) [inline applications/nfc/helpers/nfc_generators.c]: No need in this function. Please make one common function:
nfc_generate_mf_classic(NfcDeviceData* data, uint8_t uid_len, MfClassicType type) {
    uint8_t total_sectors = mf_classic_get_total_sectors_num(type)
   ...

**Review by gornekich — CHANGES_REQUESTED** [review]

**Review by gornekich — APPROVED** [review]
Good work, thanks!

**Review by DrZlo13 — APPROVED** [review]

**Review by skotopes — APPROVED** [review]

**Pull request merged**